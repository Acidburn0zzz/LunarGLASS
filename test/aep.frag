#version 310 es

precision highp float;

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_KHR_blend_equation_advanced
// + GL_OES_sample_variables
// - GL_OES_shader_multisample_interpolation
// + GL_EXT_geometry_shader
// + GL_EXT_shader_io_blocks

in inName {
    vec4 color;
} inInst;

out vec4 color;

int foo_GS();
void goodSample();

void main()
{
    color = inInst.color;
    color *= float(foo_GS());
    goodSample();
}

// GL_OES_geometry_shader

highp int foo_GS()
{
    highp int l = gl_Layer;
    highp int p = gl_PrimitiveID;
    return l + p;
}


// GL_OES_sample_variables

void goodSample()
{
    lowp     int  a1 = gl_SampleID;
    mediump  vec2 a2 = gl_SamplePosition; 
    highp    int  a3 = gl_SampleMaskIn[0];
    gl_SampleMask[0] = a3;
    mediump int n1 = gl_MaxSamples;
    mediump int n2 = gl_NumSamples;
    gl_SampleMask[0] += n1 + n2 + a1;
    color.xy += a2;
}
