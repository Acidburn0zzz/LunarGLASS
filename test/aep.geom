#version 310 es

// + GL_EXT_geometry_shader
// + GL_EXT_geometry_point_size

#extension GL_ANDROID_extension_pack_es31a : enable

precision mediump float;

in fromVertex {
    in vec3 color;
} fromV[];

in vec4 nonBlockUnsized[];

out toFragment {
    out vec3 color;
} toF;

out fromVertex {
    centroid vec3 color;
    int len;
    int inv;
};

layout(triangle_strip) out;
layout(max_vertices = 200) out;
layout(lines_adjacency) in;
layout(invocations = 4) in;

centroid in vec3 centr[];

void pointSize2();

void main()
{
    EmitVertex();
    EndPrimitive();

    color = fromV[0].color + centr[1];
    gl_Position = gl_in[0].gl_Position;

    gl_PrimitiveID = gl_PrimitiveIDIn;
    gl_Layer = 2;

    len = fromV.length();     // 4: lines_adjacency
    inv = gl_InvocationID;

    pointSize2();
}


#extension GL_OES_geometry_point_size : enable

void pointSize2()
{
    highp float ps = gl_in[3].gl_PointSize;
    gl_PointSize = ps;
}
