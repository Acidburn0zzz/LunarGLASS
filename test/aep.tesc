#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// GL_EXT_tessellation_shader: patch
// GL_EXT_tessellation_point_size
// GL_OES_primitive_bounding_box

layout(vertices = 4) out;
int outa[gl_out.length()];

patch out vec4 patchOut;

void bb();
void pointSize();
float foop();

void main()
{
    barrier();

    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    int iid = gl_InvocationID;

    gl_out[1].gl_Position = p;

    gl_TessLevelOuter[3] = 3.2;
    gl_TessLevelInner[1] = 1.3;
    bb();
    pointSize();
    patchOut = vec4(0.5);
    patchOut *= foop();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

#extension GL_ARB_separate_shader_objects : enable

layout(location = 3) in vec4 ivla[];
layout(location = 4) in vec4 ivlb[];

layout(location = 3) out vec4 ovla[];
layout(location = 4) out vec4 ovlb[];

patch out pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2[];
centroid in vec3 centr[];
out float okaySize[4];

#extension GL_OES_tessellation_point_size : enable

// GL_EXT_tessellation_point_size

void pointSize()
{
    gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
}

// for testing with gpu_shader5
in vec3 inv[];
precise vec3 pv3;

float foop()
{
    precise float d;

    pv3 = inv[3];
    pv3 *= pv3;
    pv3 = fma(pv3, pv3, pv3);
    d = fma(pv3.x, pv3.y, pv3.z);

    return d;
}

// GL_OES_primitive_bounding_box

void bb()
{
    gl_BoundingBoxOES[0] = vec4(0.0);
    gl_BoundingBoxOES[1] = vec4(1.0);
}
