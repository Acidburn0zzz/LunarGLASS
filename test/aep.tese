#version 310 es

#extension GL_ANDROID_extension_pack_es31a : enable

// - GL_EXT_tessellation_shader
// + GL_EXT_tessellation_point_size

layout(quads, ccw) in;  // - no ccw

layout(fractional_odd_spacing) in;    

layout(point_mode) in;

patch in vec4 patchIn;

void pointSize();

void main()
{
    vec4 p = gl_in[1].gl_Position;

    int pvi = gl_PatchVerticesIn;
    int pid = gl_PrimitiveID;
    vec3 tc = gl_TessCoord;
    float tlo = gl_TessLevelOuter[3];
    float tli = gl_TessLevelInner[1];

    gl_Position = p;
    pointSize();
}

in vec2 inb[];
in vec2 ind[gl_MaxPatchVertices];

in testblb {
    int f;
} blb[];

in testbld {
    int f;
} bld[gl_MaxPatchVertices];

layout(location = 23) in vec4 ivla[];
layout(location = 24) in vec4 ivlb[];

layout(location = 23) out vec4 ovla[2];

patch in pinbn {
    int a;
} pinbi;

centroid out vec3 myColor2;
centroid in vec3 centr[];

#extension GL_OES_tessellation_point_size : enable

void pointSize()
{
    gl_PointSize = gl_in[1].gl_PointSize;
}
