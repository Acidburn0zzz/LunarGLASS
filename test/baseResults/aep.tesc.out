
Top IR:
; ModuleID = 'Glslang'

%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.0 = type { <4 x float>, float }
%pinbn = type { i32 }

@gl_in = global [32 x %gl_PerVertex] zeroinitializer
@gl_PatchVerticesIn = global i32 0
@gl_PrimitiveID = global i32 0
@gl_InvocationID = global i32 0
@gl_out = global [4 x %gl_PerVertex.0] zeroinitializer
@gl_TessLevelOuter = global [4 x float] zeroinitializer
@gl_TessLevelInner = global [2 x float] zeroinitializer
@patchOut = global <4 x float> zeroinitializer
@inv = global [32 x <3 x float>] zeroinitializer
@gl_BoundingBoxOES = global [2 x <4 x float>] zeroinitializer
@outa = internal global [4 x i32] zeroinitializer
@inb = global [32 x <2 x float>] zeroinitializer
@ind = global [32 x <2 x float>] zeroinitializer
@ivla = global [32 x <4 x float>] zeroinitializer
@ivlb = global [32 x <4 x float>] zeroinitializer
@ovla = global [4 x <4 x float>] zeroinitializer
@ovlb = global [4 x <4 x float>] zeroinitializer
@pinbi = global %pinbn zeroinitializer
@myColor2 = global [4 x <3 x float>] zeroinitializer
@centr = global [32 x <3 x float>] zeroinitializer
@okaySize = global [4 x float] zeroinitializer

define fastcc void @main() {
entry:
  %iid = alloca i32
  %pid = alloca i32
  %pvi = alloca i32
  %p = alloca <4 x float>
  br label %mainBody

mainBody:                                         ; preds = %entry
  call void @llvm.gla.barrier()
  %p1 = load <4 x float>* getelementptr inbounds ([32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 0)
  store <4 x float> %p1, <4 x float>* %p
  %pvi2 = load i32* @gl_PatchVerticesIn
  store i32 %pvi2, i32* %pvi
  %pid3 = load i32* @gl_PrimitiveID
  store i32 %pid3, i32* %pid
  %iid4 = load i32* @gl_InvocationID
  store i32 %iid4, i32* %iid
  %0 = load <4 x float>* %p
  store <4 x float> %0, <4 x float>* getelementptr inbounds ([4 x %gl_PerVertex.0]* @gl_out, i32 0, i32 1, i32 0)
  store float 0x40099999A0000000, float* getelementptr inbounds ([4 x float]* @gl_TessLevelOuter, i32 0, i32 3)
  store float 0x3FF4CCCCC0000000, float* getelementptr inbounds ([2 x float]* @gl_TessLevelInner, i32 0, i32 1)
  call void @"bb("()
  call void @"pointSize("()
  store <4 x float> <float 5.000000e-01, float 5.000000e-01, float 5.000000e-01, float 5.000000e-01>, <4 x float>* @patchOut
  %1 = call float @"foop("()
  %2 = load <4 x float>* @patchOut
  %3 = insertelement <4 x float> undef, float %1, i32 0, !gla.precision !50
  %4 = insertelement <4 x float> %3, float %1, i32 1, !gla.precision !50
  %5 = insertelement <4 x float> %4, float %1, i32 2, !gla.precision !50
  %6 = insertelement <4 x float> %5, float %1, i32 3, !gla.precision !50
  %patchOut = fmul <4 x float> %2, %6, !gla.precision !50
  store <4 x float> %patchOut, <4 x float>* @patchOut
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %mainBody
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc void @"pointSize("() #0 {
entry:
  %0 = load float* getelementptr inbounds ([32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 1)
  store float %0, float* getelementptr inbounds ([4 x %gl_PerVertex.0]* @gl_out, i32 0, i32 1, i32 1)
  ret void
}

; Function Attrs: alwaysinline
define internal fastcc float @"foop("() #0 {
entry:
  %d = alloca float
  %pv3 = alloca <3 x float>
  %pv31 = load <3 x float>* getelementptr inbounds ([32 x <3 x float>]* @inv, i32 0, i32 3)
  store <3 x float> %pv31, <3 x float>* %pv3
  %0 = load <3 x float>* %pv3
  %1 = load <3 x float>* %pv3
  %pv32 = fmul <3 x float> %1, %0, !gla.precision !50
  store <3 x float> %pv32, <3 x float>* %pv3
  %2 = load <3 x float>* %pv3
  %3 = load <3 x float>* %pv3
  %4 = load <3 x float>* %pv3
  %pv33 = call <3 x float> @llvm.gla.fFma.v3f32.v3f32.v3f32.v3f32(<3 x float> %2, <3 x float> %3, <3 x float> %4), !gla.precision !50
  store <3 x float> %pv33, <3 x float>* %pv3
  %5 = load <3 x float>* %pv3
  %6 = extractelement <3 x float> %5, i32 0, !gla.precision !50
  %7 = load <3 x float>* %pv3
  %8 = extractelement <3 x float> %7, i32 1, !gla.precision !50
  %9 = load <3 x float>* %pv3
  %10 = extractelement <3 x float> %9, i32 2, !gla.precision !50
  %d5 = call float @llvm.gla.fFma.f32.f32.f32.f32(float %6, float %8, float %10), !gla.precision !50
  store float %d5, float* %d
  %11 = load float* %d
  ret float %11

post-return:                                      ; No predecessors!
  unreachable
}

; Function Attrs: alwaysinline
define internal fastcc void @"bb("() #0 {
entry:
  store <4 x float> zeroinitializer, <4 x float>* getelementptr inbounds ([2 x <4 x float>]* @gl_BoundingBoxOES, i32 0, i32 0)
  store <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>, <4 x float>* getelementptr inbounds ([2 x <4 x float>]* @gl_BoundingBoxOES, i32 0, i32 1)
  ret void
}

; Function Attrs: nounwind
declare void @llvm.gla.barrier() #1

; Function Attrs: nounwind readnone
declare <3 x float> @llvm.gla.fFma.v3f32.v3f32.v3f32.v3f32(<3 x float>, <3 x float>, <3 x float>) #2

; Function Attrs: nounwind readnone
declare float @llvm.gla.fFma.f32.f32.f32.f32(float, float, float) #2

attributes #0 = { alwaysinline }
attributes #1 = { nounwind }
attributes #2 = { nounwind readnone }

!gla.numVertices = !{!0}
!gla.entrypoint = !{!1}
!gla.inputs = !{!2, !8, !10, !12, !14, !16, !18, !20, !22, !24}
!gla.outputs = !{!26, !33, !35, !37, !39, !41, !42, !43, !46, !48}
!gla.noStaticUse = !{!16, !18, !20, !22, !41, !42, !43, !46, !24, !48}

!0 = metadata !{i32 4}
!1 = metadata !{metadata !"main", i32 15}
!2 = metadata !{metadata !"gl_in", i32 17, [32 x %gl_PerVertex]* @gl_in_typeProxy, metadata !3, metadata !4}
!3 = metadata !{i32 0, i32 0, i32 1024, null, i32 0}
!4 = metadata !{metadata !"gl_PerVertex", metadata !5, metadata !"gl_Position", metadata !6, metadata !"gl_PointSize", metadata !6}
!5 = metadata !{i32 0, i32 0, i32 1024, null}
!6 = metadata !{metadata !"", metadata !7}
!7 = metadata !{i32 0, i32 3, i32 1024, null}
!8 = metadata !{metadata !"gl_PatchVerticesIn", i32 1, i32* @gl_PatchVerticesIn_typeProxy, metadata !9}
!9 = metadata !{i32 0, i32 3, i32 1088, null, i32 0, i32 34}
!10 = metadata !{metadata !"gl_PrimitiveID", i32 1, i32* @gl_PrimitiveID_typeProxy, metadata !11}
!11 = metadata !{i32 0, i32 3, i32 1089, null, i32 0, i32 31}
!12 = metadata !{metadata !"gl_InvocationID", i32 1, i32* @gl_InvocationID_typeProxy, metadata !13}
!13 = metadata !{i32 0, i32 3, i32 1090, null, i32 0, i32 30}
!14 = metadata !{metadata !"inv", i32 1, [32 x <3 x float>]* @inv_typeProxy, metadata !15}
!15 = metadata !{i32 0, i32 3, i32 1106, null, i32 0}
!16 = metadata !{metadata !"inb", i32 1, [32 x <2 x float>]* @inb_typeProxy, metadata !17}
!17 = metadata !{i32 0, i32 3, i32 1140, null, i32 0}
!18 = metadata !{metadata !"ind", i32 1, [32 x <2 x float>]* @ind_typeProxy, metadata !19}
!19 = metadata !{i32 0, i32 3, i32 1172, null, i32 0}
!20 = metadata !{metadata !"ivla", i32 1, [32 x <4 x float>]* @ivla_typeProxy, metadata !21}
!21 = metadata !{i32 0, i32 3, i32 3, null, i32 0}
!22 = metadata !{metadata !"ivlb", i32 1, [32 x <4 x float>]* @ivlb_typeProxy, metadata !23}
!23 = metadata !{i32 0, i32 3, i32 4, null, i32 0}
!24 = metadata !{metadata !"centr", i32 1, [32 x <3 x float>]* @centr_typeProxy, metadata !25}
!25 = metadata !{i32 0, i32 3, i32 1209, null, i32 512}
!26 = metadata !{metadata !"gl_out", i32 16, [4 x %gl_PerVertex.0]* @gl_out_typeProxy, metadata !27, metadata !28}
!27 = metadata !{i32 0, i32 0, i32 1091, null, i32 0}
!28 = metadata !{metadata !"gl_PerVertex", metadata !5, metadata !"gl_Position", metadata !29, metadata !"gl_PointSize", metadata !31}
!29 = metadata !{metadata !"", metadata !30}
!30 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9}
!31 = metadata !{metadata !"", metadata !32}
!32 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10}
!33 = metadata !{metadata !"gl_TessLevelOuter", i32 7, [4 x float]* @gl_TessLevelOuter_typeProxy, metadata !34}
!34 = metadata !{i32 0, i32 3, i32 1099, null, i32 3, i32 35}
!35 = metadata !{metadata !"gl_TessLevelInner", i32 7, [2 x float]* @gl_TessLevelInner_typeProxy, metadata !36}
!36 = metadata !{i32 0, i32 3, i32 1103, null, i32 3, i32 36}
!37 = metadata !{metadata !"patchOut", i32 7, <4 x float>* @patchOut_typeProxy, metadata !38}
!38 = metadata !{i32 0, i32 3, i32 1105, null, i32 3}
!39 = metadata !{metadata !"gl_BoundingBoxOES", i32 7, [2 x <4 x float>]* @gl_BoundingBoxOES_typeProxy, metadata !40}
!40 = metadata !{i32 0, i32 3, i32 1138, null, i32 3, i32 50}
!41 = metadata !{metadata !"ovla", i32 7, [4 x <4 x float>]* @ovla_typeProxy, metadata !21}
!42 = metadata !{metadata !"ovlb", i32 7, [4 x <4 x float>]* @ovlb_typeProxy, metadata !23}
!43 = metadata !{metadata !"pinbi", i32 16, %pinbn* @pinbi_typeProxy, metadata !44, metadata !45}
!44 = metadata !{i32 0, i32 0, i32 1204, null, i32 3}
!45 = metadata !{metadata !"pinbn", metadata !5, metadata !"a", metadata !6}
!46 = metadata !{metadata !"myColor2", i32 7, [4 x <3 x float>]* @myColor2_typeProxy, metadata !47}
!47 = metadata !{i32 0, i32 3, i32 1205, null, i32 512}
!48 = metadata !{metadata !"okaySize", i32 7, [4 x float]* @okaySize_typeProxy, metadata !49}
!49 = metadata !{i32 0, i32 3, i32 1241, null, i32 0}
!50 = metadata !{i32 3}


Bottom IR:
; ModuleID = 'Glslang'

%gl_PerVertex = type { <4 x float>, float }
%gl_PerVertex.0 = type { <4 x float>, float }
%pinbn = type { i32 }

@gl_in = global [32 x %gl_PerVertex] zeroinitializer
@gl_PatchVerticesIn = global i32 0
@gl_PrimitiveID = global i32 0
@gl_InvocationID = global i32 0
@gl_out = global [4 x %gl_PerVertex.0] zeroinitializer
@gl_TessLevelOuter = global [4 x float] zeroinitializer
@gl_TessLevelInner = global [2 x float] zeroinitializer
@patchOut = global <4 x float> zeroinitializer
@inv = global [32 x <3 x float>] zeroinitializer
@gl_BoundingBoxOES = global [2 x <4 x float>] zeroinitializer
@inb = global [32 x <2 x float>] zeroinitializer
@ind = global [32 x <2 x float>] zeroinitializer
@ivla = global [32 x <4 x float>] zeroinitializer
@ivlb = global [32 x <4 x float>] zeroinitializer
@ovla = global [4 x <4 x float>] zeroinitializer
@ovlb = global [4 x <4 x float>] zeroinitializer
@pinbi = global %pinbn zeroinitializer
@myColor2 = global [4 x <3 x float>] zeroinitializer
@centr = global [32 x <3 x float>] zeroinitializer
@okaySize = global [4 x float] zeroinitializer

define fastcc void @main() {
entry:
  call void @llvm.gla.barrier()
  %gla_constGEP = getelementptr [32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 0
  %p1 = load <4 x float>* %gla_constGEP, align 16
  %gla_constGEP5 = getelementptr [4 x %gl_PerVertex.0]* @gl_out, i32 0, i32 1, i32 0
  store <4 x float> %p1, <4 x float>* %gla_constGEP5, align 16
  %gla_constGEP6 = getelementptr [4 x float]* @gl_TessLevelOuter, i32 0, i32 3
  store float 0x40099999A0000000, float* %gla_constGEP6, align 4
  %gla_constGEP7 = getelementptr [2 x float]* @gl_TessLevelInner, i32 0, i32 1
  store float 0x3FF4CCCCC0000000, float* %gla_constGEP7, align 4
  %gla_constGEP8 = getelementptr [2 x <4 x float>]* @gl_BoundingBoxOES, i32 0, i32 0
  store <4 x float> zeroinitializer, <4 x float>* %gla_constGEP8, align 16
  %gla_constGEP9 = getelementptr [2 x <4 x float>]* @gl_BoundingBoxOES, i32 0, i32 1
  store <4 x float> <float 1.000000e+00, float 1.000000e+00, float 1.000000e+00, float 1.000000e+00>, <4 x float>* %gla_constGEP9, align 16
  %gla_constGEP10 = getelementptr [32 x %gl_PerVertex]* @gl_in, i32 0, i32 1, i32 1
  %0 = load float* %gla_constGEP10, align 16
  %gla_constGEP11 = getelementptr [4 x %gl_PerVertex.0]* @gl_out, i32 0, i32 1, i32 1
  store float %0, float* %gla_constGEP11, align 16
  store <4 x float> <float 5.000000e-01, float 5.000000e-01, float 5.000000e-01, float 5.000000e-01>, <4 x float>* @patchOut, align 16
  %gla_constGEP12 = getelementptr [32 x <3 x float>]* @inv, i32 0, i32 3
  %pv31.i = load <3 x float>* %gla_constGEP12, align 16
  %pv32.i = fmul <3 x float> %pv31.i, %pv31.i, !gla.precision !50
  %pv33.i = call <3 x float> @llvm.gla.fFma.v3f32.v3f32.v3f32.v3f32(<3 x float> %pv32.i, <3 x float> %pv32.i, <3 x float> %pv32.i), !gla.precision !50
  %1 = extractelement <3 x float> %pv33.i, i32 0, !gla.precision !50
  %2 = extractelement <3 x float> %pv33.i, i32 1, !gla.precision !50
  %3 = extractelement <3 x float> %pv33.i, i32 2, !gla.precision !50
  %d5.i = call float @llvm.gla.fFma.f32.f32.f32.f32(float %1, float %2, float %3), !gla.precision !50
  %4 = call <4 x float> @llvm.gla.fSwizzle.v4f32.f32.v4i32(float %d5.i, <4 x i32> zeroinitializer)
  %patchOut = fmul <4 x float> %4, <float 5.000000e-01, float 5.000000e-01, float 5.000000e-01, float 5.000000e-01>, !gla.precision !50
  store <4 x float> %patchOut, <4 x float>* @patchOut, align 16
  br label %stage-epilogue

stage-epilogue:                                   ; preds = %entry
  br label %stage-exit

stage-exit:                                       ; preds = %stage-epilogue
  ret void
}

; Function Attrs: nounwind
declare void @llvm.gla.barrier() #0

; Function Attrs: nounwind readnone
declare <3 x float> @llvm.gla.fFma.v3f32.v3f32.v3f32.v3f32(<3 x float>, <3 x float>, <3 x float>) #1

; Function Attrs: nounwind readnone
declare float @llvm.gla.fFma.f32.f32.f32.f32(float, float, float) #1

; Function Attrs: nounwind readnone
declare <4 x float> @llvm.gla.fSwizzle.v4f32.f32.v4i32(float, <4 x i32>) #1

attributes #0 = { nounwind }
attributes #1 = { nounwind readnone }

!gla.numVertices = !{!0}
!gla.entrypoint = !{!1}
!gla.inputs = !{!2, !8, !10, !12, !14, !16, !18, !20, !22, !24}
!gla.outputs = !{!26, !33, !35, !37, !39, !41, !42, !43, !46, !48}
!gla.noStaticUse = !{!16, !18, !20, !22, !41, !42, !43, !46, !24, !48}

!0 = metadata !{i32 4}
!1 = metadata !{metadata !"main", i32 15}
!2 = metadata !{metadata !"gl_in", i32 17, [32 x %gl_PerVertex]* @gl_in_typeProxy, metadata !3, metadata !4}
!3 = metadata !{i32 0, i32 0, i32 1024, null, i32 0}
!4 = metadata !{metadata !"gl_PerVertex", metadata !5, metadata !"gl_Position", metadata !6, metadata !"gl_PointSize", metadata !6}
!5 = metadata !{i32 0, i32 0, i32 1024, null}
!6 = metadata !{metadata !"", metadata !7}
!7 = metadata !{i32 0, i32 3, i32 1024, null}
!8 = metadata !{metadata !"gl_PatchVerticesIn", i32 1, i32* @gl_PatchVerticesIn_typeProxy, metadata !9}
!9 = metadata !{i32 0, i32 3, i32 1088, null, i32 0, i32 34}
!10 = metadata !{metadata !"gl_PrimitiveID", i32 1, i32* @gl_PrimitiveID_typeProxy, metadata !11}
!11 = metadata !{i32 0, i32 3, i32 1089, null, i32 0, i32 31}
!12 = metadata !{metadata !"gl_InvocationID", i32 1, i32* @gl_InvocationID_typeProxy, metadata !13}
!13 = metadata !{i32 0, i32 3, i32 1090, null, i32 0, i32 30}
!14 = metadata !{metadata !"inv", i32 1, [32 x <3 x float>]* @inv_typeProxy, metadata !15}
!15 = metadata !{i32 0, i32 3, i32 1106, null, i32 0}
!16 = metadata !{metadata !"inb", i32 1, [32 x <2 x float>]* @inb_typeProxy, metadata !17}
!17 = metadata !{i32 0, i32 3, i32 1140, null, i32 0}
!18 = metadata !{metadata !"ind", i32 1, [32 x <2 x float>]* @ind_typeProxy, metadata !19}
!19 = metadata !{i32 0, i32 3, i32 1172, null, i32 0}
!20 = metadata !{metadata !"ivla", i32 1, [32 x <4 x float>]* @ivla_typeProxy, metadata !21}
!21 = metadata !{i32 0, i32 3, i32 3, null, i32 0}
!22 = metadata !{metadata !"ivlb", i32 1, [32 x <4 x float>]* @ivlb_typeProxy, metadata !23}
!23 = metadata !{i32 0, i32 3, i32 4, null, i32 0}
!24 = metadata !{metadata !"centr", i32 1, [32 x <3 x float>]* @centr_typeProxy, metadata !25}
!25 = metadata !{i32 0, i32 3, i32 1209, null, i32 512}
!26 = metadata !{metadata !"gl_out", i32 16, [4 x %gl_PerVertex.0]* @gl_out_typeProxy, metadata !27, metadata !28}
!27 = metadata !{i32 0, i32 0, i32 1091, null, i32 0}
!28 = metadata !{metadata !"gl_PerVertex", metadata !5, metadata !"gl_Position", metadata !29, metadata !"gl_PointSize", metadata !31}
!29 = metadata !{metadata !"", metadata !30}
!30 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 9}
!31 = metadata !{metadata !"", metadata !32}
!32 = metadata !{i32 0, i32 3, i32 1024, null, i32 -1, i32 10}
!33 = metadata !{metadata !"gl_TessLevelOuter", i32 7, [4 x float]* @gl_TessLevelOuter_typeProxy, metadata !34}
!34 = metadata !{i32 0, i32 3, i32 1099, null, i32 3, i32 35}
!35 = metadata !{metadata !"gl_TessLevelInner", i32 7, [2 x float]* @gl_TessLevelInner_typeProxy, metadata !36}
!36 = metadata !{i32 0, i32 3, i32 1103, null, i32 3, i32 36}
!37 = metadata !{metadata !"patchOut", i32 7, <4 x float>* @patchOut_typeProxy, metadata !38}
!38 = metadata !{i32 0, i32 3, i32 1105, null, i32 3}
!39 = metadata !{metadata !"gl_BoundingBoxOES", i32 7, [2 x <4 x float>]* @gl_BoundingBoxOES_typeProxy, metadata !40}
!40 = metadata !{i32 0, i32 3, i32 1138, null, i32 3, i32 50}
!41 = metadata !{metadata !"ovla", i32 7, [4 x <4 x float>]* @ovla_typeProxy, metadata !21}
!42 = metadata !{metadata !"ovlb", i32 7, [4 x <4 x float>]* @ovlb_typeProxy, metadata !23}
!43 = metadata !{metadata !"pinbi", i32 16, %pinbn* @pinbi_typeProxy, metadata !44, metadata !45}
!44 = metadata !{i32 0, i32 0, i32 1204, null, i32 3}
!45 = metadata !{metadata !"pinbn", metadata !5, metadata !"a", metadata !6}
!46 = metadata !{metadata !"myColor2", i32 7, [4 x <3 x float>]* @myColor2_typeProxy, metadata !47}
!47 = metadata !{i32 0, i32 3, i32 1205, null, i32 512}
!48 = metadata !{metadata !"okaySize", i32 7, [4 x float]* @okaySize_typeProxy, metadata !49}
!49 = metadata !{i32 0, i32 3, i32 1241, null, i32 0}
!50 = metadata !{i32 3}
#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_ARB_separate_shader_objects : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_tessellation_point_size : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(vertices = 4) out;
in gl_PerVertex {
	highp vec4 gl_Position;
	highp float gl_PointSize;
} gl_in[32];
in highp vec3 inv[32];
in highp vec2 inb[32];
in highp vec2 ind[32];
layout(location=3) in highp vec4 ivla[32];
layout(location=4) in highp vec4 ivlb[32];
in highp vec3 centr[32];
out gl_PerVertex {
	highp vec4 gl_Position;
	highp float gl_PointSize;
} gl_out[4];
patch out highp vec4 patchOut;
layout(location=3) out highp vec4 ovla[4];
layout(location=4) out highp vec4 ovlb[4];
patch out pinbn {
	highp int a;
} pinbi;
out highp vec3 myColor2[4];
out highp float okaySize[4];
const float C_3d2 = 3.2;
const float C_1d3 = 1.3;
const vec4 C_vec4p0d0p = vec4(0.0);
const vec4 C_vec4p1d0p = vec4(1.0);
const vec4 C_vec4p0d5p = vec4(0.5);

void main()
{
	barrier();
	gl_out[1].gl_Position = gl_in[1].gl_Position;
	gl_TessLevelOuter[3] = C_3d2;
	gl_TessLevelInner[1] = C_1d3;
	gl_BoundingBoxOES[0] = C_vec4p0d0p;
	gl_BoundingBoxOES[1] = C_vec4p1d0p;
	gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
	patchOut = C_vec4p0d5p;
	highp vec3 pv = inv[3] * inv[3];
	highp vec3 pv1 = fma(pv, pv, pv);
	highp float H_0ow2u2 = fma(pv1.x, pv1.y, pv1.z);
	vec4 H_iquwvs = vec4(H_0ow2u2);
	highp vec4 patchOut1 = C_vec4p0d5p * H_iquwvs;
	patchOut = patchOut1;
	
}

#version 310 es
// LunarGOO output
#extension GL_ANDROID_extension_pack_es31a : enable
#extension GL_ARB_separate_shader_objects : enable
#extension GL_EXT_geometry_shader : enable
#extension GL_EXT_gpu_shader5 : enable
#extension GL_EXT_primitive_bounding_box : enable
#extension GL_EXT_shader_io_blocks : enable
#extension GL_EXT_tessellation_shader : enable
#extension GL_EXT_texture_buffer : enable
#extension GL_EXT_texture_cube_map_array : enable
#extension GL_KHR_blend_equation_advanced : enable
#extension GL_OES_sample_variables : enable
#extension GL_OES_shader_image_atomic : enable
#extension GL_OES_shader_multisample_interpolation : enable
#extension GL_OES_tessellation_point_size : enable
#extension GL_OES_texture_storage_multisample_2d_array : enable
layout(vertices = 4) out;
in gl_PerVertex {
	highp vec4 gl_Position;
	highp float gl_PointSize;
} gl_in[32];
in highp vec3 inv[32];
in highp vec2 inb[32];
in highp vec2 ind[32];
layout(location=3) in highp vec4 ivla[32];
layout(location=4) in highp vec4 ivlb[32];
in highp vec3 centr[32];
out gl_PerVertex {
	highp vec4 gl_Position;
	highp float gl_PointSize;
} gl_out[4];
patch out highp vec4 patchOut;
layout(location=3) out highp vec4 ovla[4];
layout(location=4) out highp vec4 ovlb[4];
patch out pinbn {
	highp int a;
} pinbi;
out highp vec3 myColor2[4];
out highp float okaySize[4];
const float C_3d2 = 3.2;
const float C_1d3 = 1.3;
const vec4 C_vec4p0d0p = vec4(0.0);
const vec4 C_vec4p1d0p = vec4(1.0);
const vec4 C_vec4p0d5p = vec4(0.5);

void main()
{
	barrier();
	gl_out[1].gl_Position = gl_in[1].gl_Position;
	gl_TessLevelOuter[3] = C_3d2;
	gl_TessLevelInner[1] = C_1d3;
	gl_BoundingBoxOES[0] = C_vec4p0d0p;
	gl_BoundingBoxOES[1] = C_vec4p1d0p;
	gl_out[1].gl_PointSize = gl_in[1].gl_PointSize;
	patchOut = C_vec4p0d5p;
	highp vec3 pv = inv[3] * inv[3];
	highp vec3 pv1 = fma(pv, pv, pv);
	highp float H_0ow2u = fma(pv1.x, pv1.y, pv1.z);
	vec4 H_elkc6c1 = vec4(H_0ow2u);
	highp vec4 patchOut1 = C_vec4p0d5p * H_elkc6c1;
	patchOut = patchOut1;
	
}

